### PR TITLE
Adding handling of symbolic arguments to umask

### PR DESCRIPTION
**Pull request by benofbrown**
Fairly significant change. It took me a while to get my head around some of the bitwise parts but it seems to work as I'd expect.

### CODE REVIEW

File: builtin/umask.c [hunk 1]
```diff
@@ -7,21 +7,146 @@
...
-
 static const char umask_usage[] = "usage: umask [-S] [mode]\n";
 
+enum umask_symbolic_state {
+	UMASK_WHO,
+	UMASK_PERM,
+};
+
+static mode_t umask_current_mask(void) {
+	mode_t mask;
```
Comment (emersion) [inline builtin/umask.c]: This can be declared when used, a little bit later.

File: builtin/umask.c [hunk 2]
```diff
@@ -7,21 +7,146 @@
...
+	umask(mask);
+	return mask;
+}
+
+static int umask_update_mode(mode_t *mode, char action, mode_t *perm_mask, mode_t *who_mask) {
+	switch (action) {
+	case '+':
+		*mode |= (*who_mask & *perm_mask);
+		break;
+
```
Comment (emersion) [inline builtin/umask.c]: Style: no need for these newlines, `case` statements make it clear where each branch begins and ends.

File: builtin/umask.c [hunk 3]
```diff
@@ -7,21 +7,146 @@
...
+	case '-':
+		*mode &= (0777 & ~(*who_mask & *perm_mask));
+		break;
+
+	default:
+		fprintf(stderr, "unknown action -- '%c'\n", action);
+		return EXIT_FAILURE;
+		break;
```
Comment (emersion) [inline builtin/umask.c]: No need for this `break`.

File: builtin/umask.c [hunk 4]
```diff
@@ -7,21 +7,146 @@
...
+static mode_t umask_current_mask(void) {
+	mode_t mask;
+	const mode_t default_mode = 0022;
+
+	mask = umask(default_mode);
+	umask(mask);
+	return mask;
+}
+
+static int umask_update_mode(mode_t *mode, char action, mode_t *perm_mask, mode_t *who_mask) {
```
Comment (emersion) [inline builtin/umask.c]: We generally make functions return a `bool`, true indicating success and false indicating failures, when we don't need more.

File: builtin/umask.c [hunk 5]
```diff
@@ -7,21 +7,146 @@
...
+		break;
+	}
+
+	*perm_mask = *who_mask = 0;
+	return EXIT_SUCCESS;
+}
+
+static int umask_mode(mode_t *mode, char *symbolic) {
+	mode_t tmp_mode = 0777 & ~(umask_current_mask());
+	char *c = symbolic;
```
Comment (emersion) [inline builtin/umask.c]: Style: this can be declared in the `for` loop

File: builtin/umask.c [hunk 6]
```diff
@@ -7,21 +7,146 @@
...
+	enum umask_symbolic_state state = UMASK_WHO;
+	mode_t who_mask = 0;
+	mode_t perm_mask = 0;
+	char action = '\0';
+
+
```
Comment (emersion) [inline builtin/umask.c]: Style: extra newline here

File: builtin/umask.c [hunk 7]
```diff
@@ -7,21 +7,146 @@
...
+			break;
+		}
+	}
+
+	if (state == UMASK_PERM) {
+		if (umask_update_mode(&tmp_mode, action, &perm_mask, &who_mask) != EXIT_SUCCESS) {
+			return EXIT_FAILURE;
+		}
+	} else {
+		fprintf(stderr, "Unexpected state %d\n", state);
```
Comment (emersion) [inline builtin/umask.c]: This error message could be simplified to something like "missing permissions from symbolic mode".

File: builtin/umask.c [hunk 8]
```diff
@@ -60,22 +186,26 @@ int builtin_umask(struct mrsh_state *state, int argc, char *argv[]) {
...
 
 	char *endptr;
 	mode = strtol(argv[optind], &endptr, 8);
+
 	if (*endptr != '\0') {
-		fprintf(stderr, "Invalid mode %s\n", argv[optind]);
-		fprintf(stderr, umask_usage);
-		return EXIT_FAILURE;
+		if (umask_mode(&mode, argv[optind]) != EXIT_SUCCESS) {
+			fprintf(stderr, "Invalid mode %s\n", argv[optind]);
```
Comment (emersion) [inline builtin/umask.c]: If the callee prints an error message, I don't think it's necessary to print another one here.

**Comment by benofbrown** [comment]
I've made some changes after your feedback, hopefully it's all good now. I've also been looking at implementing getopt, I should be able to make a PR on that soon.

**Comment by emersion** [comment]
Sweet, thanks!

**Pull request merged**